File: projects/go/store/store.go
package store

import (
	"sync"
)

// OperationResponse represents a response from a DAO operation
type OperationResponse struct {
	ID     string                 `json:"id"`
	Status string                 `json:"status"`
	Result map[string]interface{} `json:"result,omitempty"`
	Error  string                 `json:"error,omitempty"`
}

// ResultStore provides thread-safe in-memory storage for operation results
type ResultStore struct {
	mu    sync.RWMutex
	store map[string]OperationResponse
}

// NewResultStore creates a new ResultStore instance
func NewResultStore() *ResultStore {
	return &ResultStore{
		store: make(map[string]OperationResponse),
	}
}

// Set stores an operation response by UUID
func (rs *ResultStore) Set(id string, response OperationResponse) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	rs.store[id] = response
}

// Get retrieves an operation response by UUID
func (rs *ResultStore) Get(id string) (OperationResponse, bool) {
	rs.mu.RLock()
	defer rs.mu.RUnlock()
	response, exists := rs.store[id]
	return response, exists
}

// Delete removes an operation response by UUID
func (rs *ResultStore) Delete(id string) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	delete(rs.store, id)
}

// Exists checks if a result exists for the given UUID
func (rs *ResultStore) Exists(id string) bool {
	rs.mu.RLock()
	defer rs.mu.RUnlock()
	_, exists := rs.store[id]
	return exists
}

File: projects/rust/src/store.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use serde::{Deserialize, Serialize};

/// OperationResponse represents a response from a DAO operation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OperationResponse {
    pub id: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// ResultStore provides thread-safe in-memory storage for operation results
#[derive(Clone)]
pub struct ResultStore {
    store: Arc<Mutex<HashMap<String, OperationResponse>>>,
}

impl ResultStore {
    /// Create a new ResultStore instance
    pub fn new() -> Self {
        ResultStore {
            store: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Store an operation response by UUID
    pub fn set(&self, id: String, response: OperationResponse) {
        let mut store = self.store.lock().unwrap();
        store.insert(id, response);
    }

    /// Retrieve an operation response by UUID
    pub fn get(&self, id: &str) -> Option<OperationResponse> {
        let store = self.store.lock().unwrap();
        store.get(id).cloned()
    }

    /// Remove an operation response by UUID
    pub fn delete(&self, id: &str) -> Option<OperationResponse> {
        let mut store = self.store.lock().unwrap();
        store.remove(id)
    }

    /// Check if a result exists for the given UUID
    pub fn exists(&self, id: &str) -> bool {
        let store = self.store.lock().unwrap();
        store.contains_key(id)
    }
}

impl Default for ResultStore {
    fn default() -> Self {
        Self::new()
    }
}
